docs/api_reference/src/quantities.tex

```diff
@@ -109,158 +599,4996 @@
+\rSec2[qty.mag.concepts]{Concepts}
+
+\begin{itemdecl}
+template<typename T>
+concept @\deflibconcept{MagConstant}@ = @\exposconceptnc{SymbolicConstant}@<T> && @\exposidnc{is-derived-from-specialization-of}@<T, mag_constant>();
+
+template<typename T>
+concept @\deflibconcept{Magnitude}@ = (@\exposidnc{is-specialization-of}@<T, magnitude>());
+
+template<typename T>
+concept @\defexposconceptnc{MagArg}@ = std::@\stdconcept{integral}@<T> || @\libconcept{MagConstant}@<T>;
+\end{itemdecl}
+
+\rSec2[qty.mag.types]{Types}
+
+\begin{codeblock}
+namespace mp_units {
+
+template<symbol_text Symbol, long double Value>
+  requires(Value > 0)
+struct @\libglobal{mag_constant}@ {
+  static constexpr auto @\exposidnc{symbol}@ = Symbol;       // \expos
+  static constexpr long double @\exposidnc{value}@ = Value;  // \expos
+};
+
+}
+\end{codeblock}
+
+A specialization of \tcode{mag_constant} represents a real number\irefiev{102-02-05}.
+\tcode{Symbol} is its symbol, and
+\tcode{Value} is (an approximation of) its value.
+
+\begin{codeblock}
+namespace mp_units {
+
+template<auto... Ms>
+struct @\libglobal{magnitude}@ {
+  friend consteval @\libconcept{Magnitude}@ auto operator*(magnitude lhs, @\libconcept{Magnitude}@ auto rhs);
+
+  friend consteval auto operator/(magnitude lhs, @\libconcept{Magnitude}@ auto rhs)
+  {
+    return lhs * @\exposid{pow}@<-1>(rhs);
+  }
+
+  template<int Num, int Den = 1>
+  friend consteval auto @\exposidnc{pow}@(magnitude);
+
+  template<@\libconcept{Magnitude}@ Rhs>
+  friend consteval bool operator==(magnitude, Rhs)
+  {
+    return std::is_same_v<magnitude, Rhs>;
+  }
+
+  friend consteval bool @\exposidnc{is-positive-integral-power}@(magnitude);
+
+  template<auto... Ms2>
+  friend consteval auto @\exposidnc{common-magnitude}@(magnitude, magnitude<Ms2...>);
+};
+
+}
+\end{codeblock}
+
+\pnum
+A specialization of \tcode{magnitude}
+represents the product of its template arguments.
+A program that instantiates a specialization of \tcode{magnitude}
+that is not a possible result of the library specifications
+is ill-formed, no diagnostic required.
+
+\pnum
+For the purposes of specifying the implementation-defined limits,
+let the representation of the terms of \tcode{magnitude} be the structure
+\begin{codeblock}
+struct {
+  @\exposidnc{ratio}@ exp;
+  @\exposidnc{base-type}@ base;
+};
+\end{codeblock}
+representing the number $\tcode{base}^\tcode{exp}$,
+where \exposid{base-type} is \tcode{std::intmax_t} or a model of \libconcept{MagConstant}.
+\begin{itemize}
+\item
+There is a single term for each \exposid{base-type}.
+\item
+\tcode{exp.num} is not expanded into base.
+\begin{note}
+$2^3 = 8$ is not permitted.
+\end{note}
+\item
+\tcode{exp.den} can reduce the base.
+\begin{note}
+$4^{1/2} = 2$ is permitted.
+\end{note}
+\item
+If the result of an operation on \tcode{std::intmax_t} values is outside its range,
+the behavior is
+\impldef{behavior of \tcode{magnitude} operations that do not fit in a \tcode{std::intmax_t}}.
```

Comment:
As an implementation, how would mp-units define this behavior?
Here's my take on it.
- There are terms for powers of each prime number that represents the magnitude.
  [ _Note_:
  It doesn't matter that above reads "[t]here is a single term for each _`base-type`_".
  `mag<INTMAX_MAX> * mag<INTMAX_MAX>` is implementation-defined.
  But `mag<INTMAX_MAX>` having more than one term with _`base-type`_ = `std::intmax_t` is unobservable.
  -- _end note_ ].
- `mag_power<2, INTMAX_MAX> * mag_power<2, 1>` naturally fails
  (although `mag_power<2, INTMAX_MAX> * mag_power<3, INTMAX_MAX>` doesn't).
- Big bases just run into the limits of `constexpr` evaluation
  (although there's work going around to push this limit).